### PR TITLE
wxGUI/history: add pop-up command menu with an item for delete command from history

### PR DESCRIPTION
**Pull request by tmszi**
Add pop-up command menu with an item for delete command from history.

wxgui_add_popup_command_menu_with_delete_cmd_item

### CODE REVIEW

File: gui/wxpython/history/browser.py [hunk 1]
```diff
@@ -100,11 +122,47 @@ def UpdateHistoryModelByCommand(self, cmd):
...
+        question = _("Do you really want to delete <{}> command?").format(cmd)
+        if self._confirmDialog(question, title=_("Delete command")) == wx.ID_YES:
+            history_path = get_current_mapset_gui_history_path()
+            if history_path:
+                del_line_number = self._model.model.GetIndexOfNode(
+                    self._selected_cmd_node,
+                )[0]
+                self.showNotification.emit(message=_("Deleting <{}>").format(cmd))
+                try:
+                    with open(history_path, "r+") as f:
```
[the remark this answers is not in the document]
Comment (tmszi) [reply gui/wxpython/history/browser.py]: I refactored `update_history()` func to add or delete command from the history file.

File: gui/wxpython/history/browser.py [hunk 2]
```diff
@@ -100,11 +122,47 @@ def UpdateHistoryModelByCommand(self, cmd):
...
         self._refreshTree()
 
+    def OnDeleteCmd(self, event):
+        """Delete cmd from the history file"""
+        cmd = self._selected_cmd_node.data["command"]
+        question = _("Do you really want to delete <{}> command?").format(cmd)
+        if self._confirmDialog(question, title=_("Delete command")) == wx.ID_YES:
+            history_path = get_current_mapset_gui_history_path()
+            if history_path:
+                del_line_number = self._model.model.GetIndexOfNode(
```
[the remark this answers is not in the document]
Comment (tmszi) [reply gui/wxpython/history/browser.py]: I added `DeleteEntryFromHistory()` method.

File: gui/wxpython/history/browser.py [hunk 3]
```diff
@@ -100,11 +122,47 @@ def UpdateHistoryModelByCommand(self, cmd):
...
+                                f.write(line)
+                except OSError as e:
+                    GError(
+                        parent=self,
+                        message=str(e),
+                        caption=_("Cannot update history file"),
+                        showTraceback=False,
+                    )
+                    return
+            self._model.CreateModel()
```
[the remark this answers is not in the document]
Comment (tmszi) [reply gui/wxpython/history/browser.py]: I refactored `UpdateModel()` method to add or delete tree command node model.

File: gui/wxpython/history/tree.py [hunk 4]
```diff
@@ -20,29 +20,61 @@
...
+        """
+        if update == "add":
+            data = {"command": label}
+            self.model.AppendNode(
+                parent=self.model.root,
+                label=data["command"],
+                data=data,
+            )
+        else:
+            self.model.RemoveNode(node)
```
Comment (tmszi) [inline gui/wxpython/history/tree.py:80]: During  my testing I investigate that `RemoveNode()` method of `self.model = TreeModel(ModuleNode)` instance not working as expected, node is not removed.

https://github.com/OSGeo/grass/blob/44e7ebc46dd83dbb48e9e469a6aa7b0d5a12645c/gui/wxpython/core/treemodel.py#L142-L148

Comment (lindakladivova) [reply gui/wxpython/history/tree.py:80]: Look at the comment below.

**Comment by lindakladivova** [comment]
@tmszi I have created some changes and posted them as the PR into your fork. 

I divided the logic of Add and Delete commands as I was also thinking that the Add signal will be much more complex in the future, so it makes sense to have it separately.

The problem with RemoveNode was that the node of the Tree differs from the node of the Model. It takes me quite a lot of time to figure that out - quite a betrayal. :-)

I also added the delete signal. Except for editing the history log file, it also updates the cmd buffer (that thing was missing in the original code).

Please, could I ask you to test it?

**Comment by tmszi** [comment]
Closed by better implementation PR https://github.com/OSGeo/grass/pull/3342.

**Pull request closed without merging**